File: routers/v1/clan.py
import coc
import pendulum as pend

from bson.objectid import ObjectId
from fastapi import  Request, Response, HTTPException
from fastapi import APIRouter
from fastapi_cache.decorator import cache
from models.clan import JoinLeaveList
from slowapi import Limiter
from slowapi.util import get_ipaddr
from utils.utils import fix_tag, leagues, db_client


router = APIRouter(tags=["Clan Endpoints"])



@router.get("/clan/{clan_tag}/basic",
         name="Basic Clan Object")
@cache(expire=300)
async def clan_basic(clan_tag: str, request: Request, response: Response):
    clan_tag = fix_tag(clan_tag)
    result = await db_client.basic_clan.find_one({"tag": clan_tag})
    if result is not None:
        del result["_id"]
    return result




@router.get(
        path="/clan/{clan_tag}/join-leave",
        name="Join Leaves in a season",
        response_model=JoinLeaveList)
@cache(expire=300)
async def clan_join_leave(clan_tag: str, request: Request, response: Response, timestamp_start: int = 0, time_stamp_end: int = 9999999999, limit: int = 250):
    clan_tag = fix_tag(clan_tag)
    result = await db_client.join_leave_history.find(
        {"$and" : [
            {"clan" : clan_tag},
            {"time" : {"$gte" : pend.from_timestamp(timestamp=timestamp_start, tz=pend.UTC)}},
            {"time": {"$lte": pend.from_timestamp(timestamp=time_stamp_end, tz=pend.UTC)}}
        ]
    }, {"_id" : 0}).sort({"time" : -1}).limit(limit=limit).to_list(length=None)
    return {"items" : result}




@router.get("/clan/search",
         name="Search Clans by Filtering")
@cache(expire=300)
async def clan_filter(request: Request, response: Response,  limit: int= 100, location_id: int = None, minMembers: int = None, maxMembers: int = None,
                      minLevel: int = None, maxLevel: int = None, openType: str = None,
                      minWarWinStreak: int = None, minWarWins: int = None, minClanTrophies: int = None, maxClanTrophies: int = None, capitalLeague: str= None,
                      warLeague: str= None, memberList: bool = True, before:str =None, after: str=None):
    queries = {}
    queries['$and'] = []
    if location_id:
        queries['$and'].append({'location.id': location_id})

    if minMembers:
        queries['$and'].append({"members": {"$gte" : minMembers}})

    if maxMembers:
        queries['$and'].append({"members": {"$lte" : maxMembers}})

    if minLevel:
        queries['$and'].append({"level": {"$gte" : minLevel}})

    if maxLevel:
        queries['$and'].append({"level": {"$lte" : maxLevel}})

    if openType:
        queries['$and'].append({"type": openType})

    if capitalLeague:
        queries['$and'].append({"capitalLeague": capitalLeague})

    if warLeague:
        queries['$and'].append({"warLeague": warLeague})

    if minWarWinStreak:
        queries['$and'].append({"warWinStreak": {"$gte": minWarWinStreak}})

    if minWarWins:
        queries['$and'].append({"warWins": {"$gte": minWarWins}})

    if minClanTrophies:
        queries['$and'].append({"clanPoints": {"$gte": minClanTrophies}})

    if maxClanTrophies:
        queries['$and'].append({"clanPoints": {"$gte": maxClanTrophies}})

    if after:
        queries['$and'].append({"_id": {"$gt": ObjectId(after)}})

    if before:
        queries['$and'].append({"_id": {"$lt": ObjectId(before)}})


    if queries["$and"] == []:
        queries = {}

    limit = min(limit, 1000)
    results = await db_client.basic_clan.find(queries).limit(limit).sort("_id", 1).to_list(length=limit)
    return_data = {"items" : [], "before": "", "after" : ""}
    if results:
        return_data["before"] = str(results[0].get("_id"))
        return_data["after"] = str(results[-1].get("_id"))
        for data in results:
            del data["_id"]
            if not memberList:
                del data["memberList"]
        return_data["items"] = results
    return return_data




@router.get("/clan/{clan_tag}/historical",
         name="Historical data for a clan of player events")
@cache(expire=300)
async def clan_historical(clan_tag: str, request: Request, response: Response, timestamp_start: int = 0, time_stamp_end: int = 9999999999, limit: int = 100):
    clan_tag = fix_tag(clan_tag)

    historical_data = await db_client.player_history.find(
        {"$and": [
            {"clan": clan_tag},
            {"time": {"$gte": int(pend.from_timestamp(timestamp=timestamp_start, tz=pend.UTC).timestamp())}},
            {"time": {"$lte": int(pend.from_timestamp(timestamp=time_stamp_end, tz=pend.UTC).timestamp())}}
        ]
        }, {"_id": 0}).sort({"time": -1}).limit(limit=limit).to_list(length=25000)

    return {"items" : historical_data}

File: internal/routes/v1/clan.go
package v1

import (
	"strconv"
	"strings"

	apptypes "github.com/ClashKingInc/ClashKingAPI/internal/utils"
	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// clanBasic godoc
// @Summary Basic Clan Object
// @Description Returns the cached clan object for a clan tag.
// @Tags Clan Endpoints
// @Produce json
// @Param clan_tag path string true "Clan tag"
// @Success 200 {object} map[string]interface{}
// @Failure 500 {object} map[string]interface{}
// @Router /clan/{clan_tag}/basic [get]
func clanBasic(a apptypes.Deps) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var row bson.M
		err := a.Store.DB.Looper.Collection("clan_tags").FindOne(c.UserContext(), bson.M{"tag": clanFixTag(c.Params("clan_tag"))}).Decode(&row)
		if err != nil {
			return apptypes.JSON(c, fiber.StatusOK, nil)
		}
		return apptypes.JSON(c, fiber.StatusOK, clanWithoutID(row))
	}
}

// clanJoinLeave godoc
// @Summary Join leaves in a season
// @Description Returns join and leave history for a clan tag.
// @Tags Clan Endpoints
// @Produce json
// @Param clan_tag path string true "Clan tag"
// @Param timestamp_start query int false "Start timestamp"
// @Param time_stamp_end query int false "End timestamp"
// @Success 200 {object} map[string]interface{}
// @Failure 500 {object} map[string]interface{}
// @Router /clan/{clan_tag}/join-leave [get]
func clanJoinLeave(a apptypes.Deps) fiber.Handler {
	return func(c *fiber.Ctx) error {
		filter := bson.M{
			"clan": clanFixTag(c.Params("clan_tag")),
			"time": bson.M{
				"$gte": clanParseInt64Default(c.Query("timestamp_start"), 0),
				"$lte": clanParseInt64Default(c.Query("time_stamp_end"), 9999999999),
			},
		}
		cur, err := a.Store.DB.Looper.Collection("join_leave_history").Find(c.UserContext(), filter)
		if err != nil {
			return err
		}
		var rows []bson.M
		if err := cur.All(c.UserContext(), &rows); err != nil {
			return err
		}
		return apptypes.JSON(c, fiber.StatusOK, map[string]any{"items": clanStripIDs(rows)})
	}
}

// clanSearch godoc
// @Summary Search clans by filtering
// @Description Returns clans filtered by location, membership, and limit options.
// @Tags Clan Endpoints
// @Produce json
// @Param location_id query int false "Location ID"
// @Param limit query int false "Maximum number of results"
// @Param member_list query bool false "Include member list"
// @Success 200 {object} map[string]interface{}
// @Failure 500 {object} map[string]interface{}
// @Router /clan/search [get]
func clanSearch(a apptypes.Deps) fiber.Handler {
	return func(c *fiber.Ctx) error {
		filter := bson.M{}
		if locationID := c.Query("location_id"); locationID != "" {
			filter["location.id"] = clanParseIntDefault(locationID, 0)
		}
		cur, err := a.Store.DB.Looper.Collection("clan_tags").Find(c.UserContext(), filter)
		if err != nil {
			return err
		}
		var rows []bson.M
		if err := cur.All(c.UserContext(), &rows); err != nil {
			return err
		}
		limit := clanParseIntDefault(c.Query("limit"), 100)
		if len(rows) > limit {
			rows = rows[:limit]
		}
		memberList, _ := apptypes.QueryBool(c, "member_list", true)
		items := make([]bson.M, 0, len(rows))
		for _, row := range rows {
			clean := clanWithoutID(row)
			if !memberList {
				delete(clean, "memberList")
			}
			items = append(items, clean)
		}
		resp := map[string]any{"items": items, "before": "", "after": ""}
		if len(rows) > 0 {
			resp["before"] = rows[0]["_id"]
			resp["after"] = rows[len(rows)-1]["_id"]
		}
		return apptypes.JSON(c, fiber.StatusOK, resp)
	}
}

// clanHistorical godoc
// @Summary Historical data for a clan
// @Description Returns historical player data for a clan tag.
// @Tags Clan Endpoints
// @Produce json
// @Param clan_tag path string true "Clan tag"
// @Param timestamp_start query int false "Start timestamp"
// @Param time_stamp_end query int false "End timestamp"
// @Success 200 {object} map[string]interface{}
// @Failure 500 {object} map[string]interface{}
// @Router /clan/{clan_tag}/historical [get]
func clanHistorical(a apptypes.Deps) fiber.Handler {
	return func(c *fiber.Ctx) error {
		filter := bson.M{
			"clan": clanFixTag(c.Params("clan_tag")),
			"time": bson.M{
				"$gte": clanParseInt64Default(c.Query("timestamp_start"), 0),
				"$lte": clanParseInt64Default(c.Query("time_stamp_end"), 9999999999),
			},
		}
		cur, err := a.Store.DB.NewLooper.Collection("player_history").Find(c.UserContext(), filter)
		if err != nil {
			return err
		}
		var rows []bson.M
		if err := cur.All(c.UserContext(), &rows); err != nil {
			return err
		}
		return apptypes.JSON(c, fiber.StatusOK, map[string]any{"items": clanStripIDs(rows)})
	}
}

func clanParseIntDefault(raw string, fallback int) int {
	if raw == "" {
		return fallback
	}
	value, err := strconv.Atoi(raw)
	if err != nil {
		return fallback
	}
	return value
}

func clanParseInt64Default(raw string, fallback int64) int64 {
	if raw == "" {
		return fallback
	}
	value, err := strconv.ParseInt(raw, 10, 64)
	if err != nil {
		return fallback
	}
	return value
}

func clanFixTag(tag string) string {
	tag = strings.TrimSpace(strings.ToUpper(tag))
	tag = strings.TrimPrefix(tag, "#")
	if tag == "" {
		return ""
	}
	return "#" + tag
}

func clanStripIDs(rows []bson.M) []bson.M {
	out := make([]bson.M, 0, len(rows))
	for _, row := range rows {
		out = append(out, clanWithoutID(row))
	}
	return out
}

func clanWithoutID(doc bson.M) bson.M {
	clean := bson.M{}
	for key, value := range doc {
		if key == "_id" {
			continue
		}
		clean[key] = value
	}
	return clean
}
